src/language-js/printer-estree.js

```diff
@@ -3094,12 +3094,29 @@ function printPathNoParens(path, options, print, args) {
       return "unknown";
     case "TSVoidKeyword":
       return "void";
-    case "TSAsExpression":
-      return concat([
-        path.call(print, "expression"),
-        " as ",
-        path.call(print, "typeAnnotation")
+    case "TSAsExpression": {
+      const typeText = options.originalText.slice(
+        options.locStart(n.typeAnnotation),
+        options.locEnd(n.typeAnnotation)
+      );
+      // Defer to expansion logic for objects, arrays, and annotations wrapped in parentheses.
+      const mightBeMultiline = /^[({[]/.test(typeText.trim());
+
+      return conditionalGroup([
```

Comment:
Is this the most efficient way to do this? I was trying to get it to work with one group, but I started getting line breaks on lines that were shorter than the `printWidth`.